### PR TITLE
Fixes #30417 - add pf4 EmptyState component

### PR DESCRIPTION
**Pull request by rabajaj0509**
This PR is a continuing effort for https://github.com/theforeman/foreman_discovery/pull/503. Opening it as a draft, need to work on the secondary actions and font sizes.

PF3 EmptyState:
Pf3

PF4 EmptyState:
partitionTable

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js [hunk 1]
```diff
@@ -14,19 +15,21 @@ const DefaultEmptyState = props => {
     documentation: {
       url,
       label = __('For more information please see'),
-      buttonLabel = __('Documentation'),
+      buttonLabel = __('documentation'),
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js]: oh, why did you rename it? :)

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js [hunk 2]
```diff
@@ -12,18 +22,22 @@ const EmptyStatePattern = props => {
     header,
     description,
   } = props;
+  const iconName = `${upperFirst(icon)}Icon`;
+  const iconClass = pficons[iconName];
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: Awesome work, making this generic,
since it's a component can you rename it into upperCase?
```suggestion
  const Icon = pficons[iconName];
```

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js [hunk 3]
```diff
@@ -12,18 +22,22 @@ const EmptyStatePattern = props => {
...
-    </PfEmptyState>
+    <EmptyState variant={EmptyStateVariant.xl}>
+      <EmptyStateIcon icon={iconClass} />
+      <Title headingLevel="h5" size="4xl">
+        {header}
+      </Title>
+      <EmptyStateBody>{description}</EmptyStateBody>
+      {action}
+      <EmptyStateSecondaryActions>
+        {documentation}
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: Looking at the old implementation,
I think the documentation should be under the description, and not as an action
something like the `PfEmptyState.Info` but the newer implementation

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js [hunk 4]
```diff
@@ -23,14 +23,19 @@ const PrimaryActionButton = ({ action }) => {
...
     {title}
   </Button>
 );
 
 // eslint-disable-next-line react/prop-types
 const onClickButton = ({ onClick, title }) => (
-  <Button onClick={onClick} bsStyle="primary" bsSize="large">
+  <Button
+    component="a"
+    inoperableEvents={onClick}
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js]: I am not sure how it works, their docs are not very informative about it
I would search how other consumers used the `onClick`

**Review by Ron-Lavi — COMMENTED** [review]
Great work @rabajaj0509 !!
the design is much better with PF4 :)
left a few comments, also notice the failing tests on travis

File: webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js (hunk 1 above)
Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js]: I thought we should get rid of the capital letter cause, it had a capital word right in the middle of the sentence.
```
For more information please see Documentation
```

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js (hunk 4 above)
Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js]: Yeah, need to look into the `onClick` part!

**Comment by johnpmitsch** [comment]
We did add an pf4 Empty state component to Katello https://github.com/Katello/katello/blob/master/webpack/components/Table/EmptyStateMessage.js

This was for the new pf4 table, which I plan to move here very soon

I'm fine with updating to use this one, but just wanted to let you know so we can stay coordinated

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js [hunk 5]
```diff
@@ -12,18 +22,22 @@ const EmptyStatePattern = props => {
...
+        {/* {secondaryActions && } */}
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: can be deleted?

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js [hunk 6]
```diff
@@ -23,14 +23,19 @@ const PrimaryActionButton = ({ action }) => {
 
 // eslint-disable-next-line react/prop-types
 const urlButton = ({ url, title }) => (
-  <Button href={url} bsStyle="primary" bsSize="large">
+  <Button component="a" href={url} target="_blank" variant="primary">
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js]: Why using `target="_blank"`? doesn't this link lead to another page in foreman?

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js (hunk 4 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js]: According to the PF4's [implementation](https://github.com/patternfly/patternfly-react/blob/master/packages/react-core/src/components/Button/Button.tsx) the default rendered `component` is basically an HTML button element, plus the `<Button>` component passed all given extra props, so this should work:
```js
<Button onclick={onClick} variant='primary'> {title} </Button>
```

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js [hunk 7]
```diff
@@ -23,14 +23,19 @@ const PrimaryActionButton = ({ action }) => {
...
   </Button>
 );
 
 // eslint-disable-next-line react/prop-types
 const onClickButton = ({ onClick, title }) => (
-  <Button onClick={onClick} bsStyle="primary" bsSize="large">
+  <Button
+    component="a"
+    inoperableEvents={onClick}
+    target="_blank"
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js]: ditto

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @rabajaj0509 ! looks great :+1: 
I left a few inline comments

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js (hunk 6 above)
Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js]: I think that is responsible to open the link in a new tab.

Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js]: indeed, but is there a reason for opening it in a different tab?  doesn't it a link to another page in foreman?

**Comment by rabajaj0509** [comment]
@laviro @amirfefer  PF4 look, let me know what you folks think:

partitionTable

File: webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js [hunk 8]
```diff
@@ -11,31 +12,18 @@ const DefaultEmptyState = props => {
...
-  ) : null;
-
   return (
     <EmptyStatePattern
       icon={icon}
       iconType={iconType}
       header={header}
       description={description}
-      documentation={documentationBlock}
+      documentation={documentation}
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js:54]: can you change the `proptype` of this prop in `EmptyStatePattern` ?
since it's no more an element but an object with a specific shape, e.g:
```js
PropTypes.shape({
  url: PropTypes.string,
  label: PropTypes.string,
  buttonLabel: PropTypes.string,
}),
```

**Review by Ron-Lavi — COMMENTED** [review]
Great work @rabajaj0509 :+1:
there are still some [tests snapshots failing](https://travis-ci.org/github/theforeman/foreman/jobs/710687497#L3144)

one test is failing because the `documentation` prop have changed

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js [hunk 9]
```diff
@@ -1,29 +1,50 @@
...
+import {
+  Title,
+  Button,
+  EmptyState,
+  EmptyStateVariant,
+  EmptyStateIcon,
+  EmptyStateBody,
+  EmptyStateSecondaryActions,
+} from '@patternfly/react-core';
+import * as pficons from '@patternfly/react-icons';
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: This turns the bundle to be more than 4 mb!
we do not need to import all the pf icons here only for one usage.
IMO we should limit it to a  few opinionated icons or  we can use dynamic import.

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @rabajaj0509 :+1: 
keep in mind that old usages for this component are broken (i.e emptyState of the statistics page)
Also, we need to address the importing of the entire pf icons which make the bundle much bigger

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js (hunk 9 above)
Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: I will make make the required changed and update the pr.

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: I think that for this PR it would be better to use the PF3 icons, since this case need to be generic..
until we will solve the dynamic loading issue or another way to load those icons efficiently,
I remember there were issues with webpack 3 and dynamic loading - maybe we should re-try when moving to webpack 4
cc @sharvit @tbrisker

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: any more thoughts about this?

Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: @amirfefer @laviro I have made changes to use pf3 icons, let me know what you think!

**Comment by rabajaj0509** [comment]
@laviro  @amirfefer do the hound errors also apply to React PRs? I thought their scope was limited to Ruby.

**Comment by Ron-Lavi** [comment]
Wow, that was not expected :)
I guess hound isn't limited to the assets/ folder... 🤣 
I think we should first decide if we want it also on webpack/ (?)
And if so we need to update its rules

Imo it's spamming the PR while we can get same errors in travis or gh actions..
Maybe we should open a discussion about it?

**Comment by tbrisker** [comment]
@rabajaj0509 please rebase on latest develop to get https://github.com/theforeman/foreman/pull/7924 so we can see that hound issue is fixed

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.test.js [hunk 10]
```diff
@@ -45,25 +45,25 @@ const emptyStatePatternFixtures = {
     icon: 'printer',
     header: 'Printers',
     description: 'Printers print a file from the computer',
-    documentation: 'This is a simple description',
+    documentation: {
+      url: 'https://example.com',
+    },
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.test.js]: can you store documentation as a const in the fixtures and reuse it all over those tests?
actually, bonus points would be to move all of those props into an object in `fixtures.js`, e.g:

```js
export const props = {
  icon,
  header,
  description,
  documentation,
}
```
where this object keys will be also declared in the same file, like `export const icon = 'printer'`.

then in the test it will be just:
```js
'should render main action when given title and url': { ...props, action: actionWithLink }
```

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js [hunk 11]
```diff
@@ -1,29 +1,49 @@
...
+    <EmptyState variant={EmptyStateVariant.xl}>
+      <span className="empty-state-icon">
+        <Icon name={icon} type={iconType} size="6x" />
+      </span>
+      <Title headingLevel="h5" size="4xl">
+        {header}
+      </Title>
+      <EmptyStateBody>
+        {description}
+        <br />
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: imo we should use some css `margin` instead of `br`

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js [hunk 12]
```diff
@@ -23,14 +23,14 @@ const PrimaryActionButton = ({ action }) => {
...
-  <Button href={url} bsStyle="primary" bsSize="large">
+  <Button component="a" href={url} variant="primary">
     {title}
   </Button>
 );
 
 // eslint-disable-next-line react/prop-types
 const onClickButton = ({ onClick, title }) => (
-  <Button onClick={onClick} bsStyle="primary" bsSize="large">
+  <Button component="a" onclick={onClick} variant="primary">
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js]: Not sure if we tried that already, maybe we can combine `onClickButton` and `urlButton`
but what if we just return the button with `onClick` and `href`, something like:
```js
<Button component="a" onclick={onClick} href={url} variant="primary">
```
`url` should defaults to `#` / `''` to redirect nowhere

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js [hunk 13]
```diff
@@ -1,29 +1,49 @@
...
+import { translate as __ } from '../../../common/I18n';
+import './EmptyState.scss';
 
 const EmptyStatePattern = props => {
   const {
-    documentation,
+    documentation: {
+      url,
+      label = __('For more information please see'),
+      buttonLabel = __('documentation'),
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: default props can be given via `defaultProps`, i.e

```js
EmptyStatePattern.propTypes = {
  documentation:  PropTypes.shape({
    url: PropTypes.string.isRequired,
    label: PropTypes.string,
    buttonLabel: PropTypes.string
  })
};

EmptyStatePattern.defaultProps = {
  documentation: {
    label: __('For more information please see'),
    buttonLabel: __('documentation'),
  },
};
```

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js [hunk 14]
```diff
@@ -1,29 +1,49 @@
...
-      <PfEmptyState.Info>{description}</PfEmptyState.Info>
-      {documentation && <PfEmptyState.Help>{documentation}</PfEmptyState.Help>}
-      {action && <PfEmptyState.Action>{action}</PfEmptyState.Action>}
-      {secondaryActions && (
-        <PfEmptyState.Action secondary>{secondaryActions}</PfEmptyState.Action>
-      )}
-    </PfEmptyState>
+    <EmptyState variant={EmptyStateVariant.xl}>
+      <span className="empty-state-icon">
+        <Icon name={icon} type={iconType} size="6x" />
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: Can you please add a `TODO` here for a reminder to use pf4 icons?

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks, @rabajaj0509! Tested and works well :+1: 
Can you please update `emptyState` story as well? it uses outdated buttons and flow
Plus, I left a few inline comments

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js (hunk 11 above)
Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: I tried using `<p>` instead of a margin, looks better now :+1:

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js (hunk 12 above)
Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js]: I was thinking more in the terms of:
```
if (action.url || action.onClick) {
  if (!action.url) {
    action.url = '';
  }
  return urlButton(action);
}

const urlButton = ({ url, onClick, title }) => (
  <Button component="a" href={url} onclick={onClick} variant="primary">
    {title}
  </Button>
);
```

What do you think?

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePrimaryActionButton.js]: I think that we could implement it directly in `defaultEmptyState.js`
current usage:
```js
action={action ? <PrimaryActionButton action={action} /> : null}
```

we could change it into something like:
```js
action={action ? <Button component="a" onclick={action.onClick} href={action.url || ''} variant="primary"> : null}
```

it would be best to pass `action` to `EmptyStatePattern` and declare it there IMO

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js (hunk 11 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: it's best not to have extra empty html elements,
we could wrap `description` with a span and add css margin to it.

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: actually we could wrap it with a `div` and than what comes after it will be shifted down anyway
as `div` is a block element

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js [hunk 15]
```diff
@@ -1,29 +1,65 @@
...
+      } = documentation;
+      return (
+        <span>
+          {label}
+          <a href={url}>{buttonLabel}</a>
+        </span>
+      );
+    }
+    return documentation;
+  };
```
Comment (rabajaj0509) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: This does not work exactly as expected. If existed documentation is passed then it throws an error saying `documentation is undefined`

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: maybe you meant to drop the `!` on lines 27,28 ?

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: also IMO it can be `documentation.hasOwnProperty('label') || documentation.hasOwnProperty('buttonLabel')`

Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: > maybe you meant to drop the `!` on lines 27,28 ?

Error:
```
Uncaught Invariant Violation: Objects are not valid as a React child (found: object with keys {url}). If you meant to render a collection of children, use an array instead.
```

Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: got it: here we were returning the object `documentation` which is causing the issue!

**Comment by rabajaj0509** [comment]
I think this PR is ready now! Let me know what you folks think :smile:

File: webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js [hunk 16]
```diff
@@ -1,43 +1,58 @@
...
 
-  const documentationBlock = url ? (
-    <React.Fragment>
-      {label}{' '}
-      <a href={url} target="_blank" rel="noopener noreferrer">
-        {buttonLabel}
-      </a>
-    </React.Fragment>
+  const actionButtonClickHandler = ({ url, onClick }) => {
+    onClick && onClick();
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js]:
```suggestion
  const actionButtonClickHandler = ({ url, onClick = noop }) => {
    onClick();
```
This way we can trigger the function with no fear

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.stories.js [hunk 17]
```diff
@@ -104,7 +101,7 @@ export const foremanEmptyState = () => {
     docObject.label = text('documentation label', 'Read documents ->');
   }
   if (customizeDocButtonLabel) {
-    docObject.buttonLabel = text('documentation button label', 'Click here');
+    docObject.ButtonLabel = text('documentation button label', 'Click here');
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.stories.js]: IMO in this case it should be reverted back to lowerCased `buttonLabel` as this is the prop name

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js [hunk 18]
```diff
@@ -1,29 +1,65 @@
...
+    icon,
     header,
     description,
   } = props;
 
+  const DocumentationBlock = () => {
+    if (!documentation) {
+      return null;
+    }
+    if (React.isValidElement(documentation)) {
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js:30]: Can you add a small comment above, e.g:
"The documentation prop can be also a customized node"

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js [hunk 19]
```diff
@@ -1,29 +1,65 @@
...
-      <PfEmptyState.Info>{description}</PfEmptyState.Info>
-      {documentation && <PfEmptyState.Help>{documentation}</PfEmptyState.Help>}
-      {action && <PfEmptyState.Action>{action}</PfEmptyState.Action>}
-      {secondaryActions && (
-        <PfEmptyState.Action secondary>{secondaryActions}</PfEmptyState.Action>
-      )}
-    </PfEmptyState>
+    <EmptyState variant={EmptyStateVariant.xl}>
+      <span className="empty-state-icon">
+        {/* TODO: Add pf4 icons */}
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: can you add the issue link here?

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js [hunk 20]
```diff
@@ -1,29 +1,65 @@
...
+        <div className="empty-state-description">{description}</div>
+        <DocumentationBlock />
+      </EmptyStateBody>
+      {action}
+      <EmptyStateSecondaryActions>
+        {secondaryActions}
+      </EmptyStateSecondaryActions>
+    </EmptyState>
   );
 };
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js:66]: Can you also add `EmptyStatePattern.defaultProps` here?

**Review by Ron-Lavi — COMMENTED** [review]
Thanks @rabajaj0509 !! great work :)
left some inline comments

@johnpmitsch can you also take a look? as you were planning to use this component?

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js (hunk 20 above)
Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js:66]: @laviro can you please check if the default props are correct?

File: webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js (hunk 16 above)
Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js]: This doesn't work in the expected manner :(

**Comment by rabajaj0509** [comment]
@johnpmitsch @laviro @amirfefer do you have any idea how I can get rid of the Jenkins error :) unable to reproduce them on my local machine :smile:

**Comment by Ron-Lavi** [comment]
@rabajaj0509 Looks like that issue was with eslint and now got resolved :)

**Comment by rabajaj0509** [comment]
I guess irrelevant tests are failing, this PR seems to be ready!

**Comment by johnpmitsch** [comment]
@rabajaj0509 I didn't do a detailed review, but the storybook and usage seems good and I will plan to work this in with my table contribution :+1:

File: webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js [hunk 21]
```diff
@@ -1,43 +1,59 @@
...
-    <React.Fragment>
-      {label}{' '}
-      <a href={url} target="_blank" rel="noopener noreferrer">
-        {buttonLabel}
-      </a>
-    </React.Fragment>
+  const actionButtonClickHandler = ({ url, onClick = noop }) => {
+    onClick();
+    if (url) {
+      window.location.href = url;
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js]: @MariaAga did you mention before that using `location.href` had issues in tests?

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.scss [hunk 22]
```diff
@@ -0,0 +1,7 @@
+.pf-c-empty-state {
+  margin-top: 100px;
+
+  .empty-state-icon {
+    font-size: xxx-large;
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.scss]: https://developer.mozilla.org/en-US/docs/Web/CSS/font-size

it seem that the `xxx-large` is not supported in all latest browser, lets use a pixel size,
or even better - patternfly's font size variable

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.test.js [hunk 23]
```diff
@@ -36,60 +22,7 @@ describe('Default Empty State', () => {
...
-    secondaryActions: (
-      <SecondaryActionButtons
-        actions={[
-          { title: 'x', url: 'some-url' },
-          { title: 'y', url: 'random-url' },
-        ]}
-      />
-    ),
-  },
+  'should render with props': props,
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.test.js:32]: I liked that you reduced the amount of this snapshots tests, looking at the props it seem they had equivalent props and almost the same snapshots :+1:

File: webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js (hunk 21 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js]: From what I remember there were 2 uses for `window.location` in the same component so mocking for both cases wasn't possible. Maybe here it can be mocked easily

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.scss (hunk 22 above)
Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.scss]: you mean pf4 global CSS variable right? https://www.patternfly.org/v4/documentation/overview/global-css-variables

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.scss]: exactly, something like `--pf-global--icon--FontSize--xl`
you can import it from the patternfly node module in the .scss file as `@import "~patternfly/sass/app";` to the specific location of that var.

Comment (rabajaj0509) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.scss]: gotcha! thanks a lot @laviro :+1:

File: webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js (hunk 21 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js]: @rabajaj0509 can you add a small test for that link?
need to mock that `location.href` method, trigger a `click` on the link,
and expect the mock method to be called with the link address

EDIT: or you can mock the `actionButtonClickHandler`

Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js]: There's a new  mechanism for pushing a new URL instead of using the `window.location` (which doesn't trigger react-router) by using `push` from `conected-react-rouer`
```js
import {useDispatch } from 'react-redux';
import { push } from 'connected-react-router';

const dispatch = useDispatch()
const actionButtonClickHandler = ({ url, onClick}) => {
  if (onClick) onClick()
  else if (url) dispatch(push(url))
}
```

with this mechanism, the react-router is responsible for rendering the page.

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks, @rabajaj0509 :+1: 
Tested with the storybook and foreman and works well
I just left one inline comment about using `window.location`

File: webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js (hunk 21 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js]: @amirfefer should we use [pushUrl](https://github.com/theforeman/foreman/blob/fbe38adc5935284479f848c2387d672e6a467179/webpack/assets/javascripts/foreman_navigation.js#L34) method?

Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/EmptyState/DefaultEmptyState.js]: It's a helper mainly for legacy scripts in order to dispatch a function indirectly (therefore it exists in the window object).
In react components IMHO it would be better to dispatch an action directly from `useDispach` hook.

**Review by Ron-Lavi — COMMENTED** [review]
In the JS tests there is the following error with the `defaultEmptyState` test: 
`"could not find react-redux context value; please ensure the component is wrapped in a <Provider>"`

This happens since you added the `useDispatch` hook which needs the `store` to exist.
In order for this to work you should wrap [the component](https://github.com/theforeman/foreman/blob/0eb76b7898b10382a8fef6c08547d6cda6f14a78/webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyState.test.js#L19) with

`<Provider store={store}><Component /></Provider>`

and don't forget to do the imports on the top for:
```js
import { Provider } from "react-redux";
import store from "../../redux";
```

**Comment by rabajaj0509** [comment]
Thanks @laviro, I think the tests should pass now, I was not much aware about the hooks so the previous comment was helpful :+1:

**Review by Ron-Lavi — DISMISSED** [review]
Great work @rabajaj0509 !!
Looks good to me, @MariaAga @amirfefer can you take a final look?

File: webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js [hunk 24]
```diff
@@ -1,32 +1,80 @@
...
 
 EmptyStatePattern.propTypes = emptyStatePatternPropTypes;
 
+EmptyStatePattern.defaultProps = {
+  icon: 'add-circle-o',
+  secondaryActions: [],
+  documentation: {
+    label: 'For more information please see ',
+    buttonLabel: 'documentation',
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/common/EmptyState/EmptyStatePattern.js]: This is set in the documentation defaults, can it be null here?

**Review by MariaAga — CHANGES_REQUESTED** [review]
Code looks good! 
just a few small issues:
In the storybook `foremanEmptyState` crashes with this PR.
Also looks like the test `'should render secondary actions'` only tests that the render doesn't fail but doesn't test anything about the secondary actions, is there a way to make the test better? @laviro any thoughts about the test?

**Comment by rabajaj0509** [comment]
@MariaAga hey, I have resolved your comments, can you re-review and give your feedback :smile:

**Review by MariaAga — APPROVED** [review]
Great!

**Comment by rabajaj0509** [comment]
[test]

**Comment by Ron-Lavi** [comment]
@amirfefer @ezr-ondrej could you take another look? :)

**Review by amirfefer — APPROVED** [review]
Thanks @rabajaj0509
Great addition :+1:

**Comment by amirfefer** [comment]
@rabajaj0509  - merging is blocked due to `Redmine issues` malfunction 
could you please just rebase/update this?

**Comment by ezr-ondrej** [comment]
Close/reopen also does the job ^_^

**Comment by amirfefer** [comment]
[test katello]

**Comment by Ron-Lavi** [comment]
Thanks @rabajaj0509 @MariaAga @ezr-ondrej @amirfefer !!!
